### PR TITLE
Rewards Distribution Smart Contract

### PR DESCRIPTION
**Pull request by boivlad**
organized hardhat-deploy
wrote Asset sample
developed RewardDistributor contract
wrote tests for RewardDistributor

Tasks:
-  https://app.clickup.com/t/344n6jm
- https://app.clickup.com/t/37wwrw8

### CODE REVIEW

File: contracts_arbitrum/deploy/deployRewardDistributor.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module.exports = async ({ getNamedAccounts, deployments }) => {
+  const { deploy } = deployments;
+  const { deployer } = await getNamedAccounts();
+  await deploy('RewardsDistributor', {
+    from: deployer,
+    args: ['0xFd086bC7CD5C481DCC9C85ebE478A1C0b69FCbb9'],
```
Comment (ruslanpalagin) [inline contracts_arbitrum/deploy/deployRewardDistributor.ts:6]: Should not it be dynamic?
For staging and dev env it's another address

File: contracts_arbitrum/deploy/deployAsset.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+module.exports = async ({ getNamedAccounts, deployments }) => {
+  const { deploy } = deployments;
+  const { deployer } = await getNamedAccounts();
+  await deploy('Asset', {
+    from: deployer,
```
Comment (ruslanpalagin) [inline contracts_arbitrum/deploy/deployAsset.ts:5]: Do we need this (deploying an empty Asset) for further cloning?

File: contracts_arbitrum/contracts/v2/Asset.sol [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "./RewardsDistributor.sol";
+import "hardhat/console.sol"; // TODO: remove
+
+
+contract Asset is ERC20, Ownable {
+  RewardsDistributor public rewardsDistributor;
+
+  constructor () ERC20("Asset", "ASSET") {
+    _mint(msg.sender, 1000000000000000000000000);
```
Comment (ruslanpalagin) [inline contracts_arbitrum/contracts/v2/Asset.sol:14]: Should not it be specified by superOracle?

File: contracts_arbitrum/contracts/v2/Asset.sol [hunk 4]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: MIT
```
Comment (ruslanpalagin) [inline contracts_arbitrum/contracts/v2/Asset.sol:1]: current version - I would say it's "3"

File: contracts_arbitrum/tests/v2/RewardDistributor.ts [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+import * as assert from 'assert';
+import { parseEmissionPoints } from '../../testUtils/objectUtils';
+
+const hre = require('hardhat');
+
+const mockData = {
+  rewardDataOne: {
+    alice: {
+      amount: ethers.utils.parseEther('25'),
+      result: ethers.utils.parseEther('13.1428571428'),
```
Comment (ruslanpalagin) [inline contracts_arbitrum/tests/v2/RewardDistributor.ts:18]: it's very non obvious what it's 13.1428571428 and what is the given:
1. Token price
2. Time elapsed
3. Is Asset sold out or rejected (another task - just not to forget)

File: contracts_arbitrum/tests/v2/RewardDistributor.ts [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+      const carolBalance = await usdtToken.balanceOf(carol.address);
+
+      const rewardDistributorBalance = await usdtToken.balanceOf(rewardDistributor.address);
+
+      assert.equal(poolInfo.totalSupply.toString(), ethers.utils.parseEther('175').toString());
+      assert.equal(poolInfo.currentEmissionPoint.toString(), '3');
+      assert.equal(aliceBalance.toString(), mockData.rewardDataOne.alice.result.toString());
+      assert.equal(bobBalance.toString(), mockData.rewardDataOne.bob.result.toString());
+      assert.equal(carolBalance.toString(), mockData.rewardDataOne.carol.result.toString());
+      assert.equal(rewardDistributorBalance.toString(), '908000000000400000000');
```
Comment (ruslanpalagin) [inline contracts_arbitrum/tests/v2/RewardDistributor.ts:143]: Can we make this somehow more descriptive? I mean why it's 908000000000400000000 and not 9008000000000400000000 for example?  How did you get this number?

**Review by ruslanpalagin — CHANGES_REQUESTED** [review]

File: contracts_arbitrum/contracts/v2/Asset.sol (hunk 3 above)
Comment (boivlad) [reply contracts_arbitrum/contracts/v2/Asset.sol:14]: This is just for the mock, before the Asset are implemented

File: contracts_arbitrum/deploy/deployAsset.ts (hunk 2 above)
Comment (boivlad) [reply contracts_arbitrum/deploy/deployAsset.ts:5]: Just now for mock

File: contracts_arbitrum/tests/v2/RewardDistributor.ts (hunk 5 above)
Comment (boivlad) [reply contracts_arbitrum/tests/v2/RewardDistributor.ts:18]: 1) Token price is not used in current cases
2) Added
3) Of course

File: contracts_arbitrum/deploy/deployRewardDistributor.ts (hunk 1 above)
Comment (boivlad) [reply contracts_arbitrum/deploy/deployRewardDistributor.ts:6]: Added config for different chaines

**Review by ruslanpalagin — APPROVED** [review]

**Pull request merged**